### ISSUE TITLE
Improve quality of socket loop

### ISSUE THREAD

**Issue by huitema** [private-octopus/picoquic#1603]
The initial design of picoquic envisaged that applications would provide their own "socket loop", but we realize now that many would prefer to run picoquic in a background thread. The current "sockloop.c" would work, but it suffers from a few issues:

* there is no way for the app to "wake up" the stack when new data is ready,
* the code is a bit too complex because it includes demonstration logic for demo scenarios, such as simulating multihoming with 2 ports.
* there are two separate loops for Windows and Linux.
* there is not much documentation.
* the time check callback should recompute the time if the application calls the API to insert data

In addition, we need to develop automated testing.

**Comment by victorstewart** [comment]
if it were Linux only, you could use io_uring and send buffers (which implemented some header + payload message scheme) back and forth between the host and the picoquic loop without any synchronization necessary. this is what i do in my multithreaded processes. all threads run nonstop, harvesting inter-thread communication intertwined in sequence with any other operating system interactions (sendmsg, recvmsg, etc)

**Comment by huitema** [comment]
Todo -- verify type and use of ifindex -- unsigned int for Linux, ULONG for Windows.

**Comment by huitema** [comment]
I needed to take a closer look at time and clocks before using the "monotonic" clock.

Picoquic uses a single clock for two reasons:

1) Computing delays between network events, such as RTT, timers, etc. 
2) Checking some variables against wall clock, e.g., checking the validity of TLS tickets, X.509 certificates, etc.

The monotonic clock is better than the wall clock for the first usage, because the wall clock can be set ahead of backward by the admin. (Both can be affected by `adjtime()` and by NTP. ) But the monotonic clock is inadequate for the second usage -- it makes no guarantee about a fixed start point in time.

For now, the simple solution may well be to keep `gettimeofday`.

**Comment by fbussery** [comment]
Just for your information and to share our experience with using picoquic:

During the phase of selecting our QUIC library, we initially overlooked the picoquic library. The main reason was that, for our project, we don't have direct access to the UDP socket. Instead, an external library provides us with a specific API that uses callbacks for UDP communication. Initially, it appeared challenging to integrate this with your socket loop implementation, and we mistakenly believed this loop was mandatory.

By the way, it is not, and I can confirm that there might be a lack of documentation on this aspect. To provide more insight into our implementation, we rely on Boost Asio to manage our own loop.

**Comment by huitema** [comment]
In some cases we observe diminished performance. The underlying infrastructure could support Gigabit/s but the application struggles sending more than 10 or 20 Mbps. The simplest explanation is that the background thread is starving. This can happen for example if the thread is waiting on a select call for a limited delay, but the select call only returns long after the delay is passed or messages have arrived.

TODO: evaluate whether the socket thread is starving. Maybe by exposing statistics on expected versus actual wait time.

**Comment by huitema** [comment]
@fbussery could you open a separate issue for the integration with an 'external' event manager? There are several possible options, maybe Boost Asio, maybe libevent, maybe something else. That's a useful topic, but it poses a set of issues different from the "socket loop in a background thread" studied here.

**Comment by fbussery** [comment]
> @fbussery could you open a separate issue for the integration with an 'external' event manager? There are several possible options, maybe Boost Asio, maybe libevent, maybe something else. That's a useful topic, but it poses a set of issues different from the "socket loop in a background thread" discussed here.

Issue opened: [https://github.com/private-octopus/picoquic/issues/1608](https://github.com/private-octopus/picoquic/issues/1608)

**Comment by huitema** [comment]
Fixed.
